### ISSUE TITLE
MAM not working after repeatedly clearing history

### ISSUE THREAD

**Issue by ChaosKid42** [siacs/Conversations#2897]
#### General information

* **Version:** 2.0.0
* **Device:** Moto G4
* **Android Version:** LineageOS 14.1 (Android 7.1.2)
* **Server name:** self-hosted
* **Server software:** ejabberd 18.01
* **Installed server modules:** Stream Managment, CSI, MAM
* **Conversations source:** self build (2.0.0)

#### Steps to reproduce

1. turn on message delivery receipts
2. write several messages in a one-to-one-chat and generate receipts
3. repeat several times until mam won't fetch messages anymore
    a) Clear history
    b) use mam to load history

#### Expected result

Mam will aways work and fetch messages from the server.

#### Actual result

Repeating the clear history / mam load loop will lead to filling up the history with message delivery receipts. After some time the last 50 (= PAGE_SIZE) messages are all message delivery receipts i.e. messages without body. Therefore Conversations seems to believe there is nothing left to fetch from the server and stops. The user cannot recover from this situation. Only recompiling with an increases value for PAGE_SIZE helps.

#### Debug output

The relevant line is:

````
03-25 14:30:29.608 D/conversations(31136): christoph@scholzbande.de: finished mam after 50(0) messages. messages left=false count=217
````

anyway here is the full log

https://gist.github.com/ChaosKid42/0cb9cece7f4e2b5c893d69ac44cca81d

**Comment by iNPUTmice** [comment]
That's a safeguard against loading worthless history.

**Comment by ChaosKid42** [comment]
I am of course not complaining about Conversations not showing message delivery receipts. I think however that the user is supposed to be able to load his personal chat history from the server in any case.
It might seem a little academic, but I myself actually triggered this problem several times while testing.
There seems to be absolutely nothing you can do about it, after you once created a block of 50 (or more) consecutive delivery receipts.
So I still believe this is worth fixing.

**Comment by iNPUTmice** [comment]
Are you sure it’s delivery receipts? To me it looks more like omemo messages. I'm not sure how one would get 50 receipts in a row into one conversation (OK I could imagine how but it wouldn’t be a naturally occurring phenomenon)

**Comment by ChaosKid42** [comment]
Yes I am sure. While testing I repeadetly cleared the history and reloaded from the archive. This seems to generate additional delivery receipts. After some time the history cannot be loaded by Conversations.

**Comment by ChaosKid42** [comment]
This is what I found by using gajim's xml console:

https://gist.github.com/ChaosKid42/523b66cae0cc172b0dc17ffdc4d3c78c

**Comment by ChaosKid42** [comment]
Looking at MessageArchiveService.java I believe the problem is that there is no MamReference saved in that case. Otherwise you could simply fetch the next page of messages from archive.

**Comment by iNPUTmice** [comment]
There is a bug that it sends receipts multiple times. Will fix that later.
The general safe guard should remain in place I believe

**Comment by ChaosKid42** [comment]
Thanks. I'm fine with that. At least then a common user wouldn't trigger this problem any more.

**Comment by iNPUTmice** [comment]
Can you provide a logcat for the 'clear history', 'load history', repeat phase? Since in that phase it supposedly adds the delivery receipt that phase seems more interesting.

**Comment by ChaosKid42** [comment]
Yes. I'll do that later today when I return home.

**Comment by licaon-kter** [comment]
@ChaosKid42 Do paste logs on https://gist.github.com and only post the link, else this makes the issue harder to load/read.

**Comment by ChaosKid42** [comment]
Here is the new logcat:

https://gist.github.com/ChaosKid42/bfaa09f8a5ece0c71a95bdcefe047c72

and after a grep mam:

https://gist.github.com/ChaosKid42/79dda93bf120b212a3ea70c95b2f1632

**Comment by iNPUTmice** [comment]
The general cause of this (repeatedly sending receipts) should be fixed in 2.1.0
